src/redis-benchmark.c

```diff
@@ -1189,85 +1226,127 @@ static int fetchClusterConfiguration() {
+                            sds dst = sdsnew(p);
+                            node->migrating_count += 2;
+                            node->migrating =
+                                zrealloc(node->migrating,
+                                    (node->migrating_count * sizeof(sds)));
+                            node->migrating[node->migrating_count - 2] =
+                                slot;
+                            node->migrating[node->migrating_count - 1] =
+                                dst;
+                        }  else if ((p = strstr(slotsdef, "-<-"))) {//Importing
+                            *p = '\0';
+                            p += 3;
+                            char *closing_bracket = strchr(p, ']');
+                            if (closing_bracket) *closing_bracket = '\0';
+                            sds slot = sdsnew(slotsdef);
+                            sds src = sdsnew(p);
+                            node->importing_count += 2;
+                            node->importing = zrealloc(node->importing,
+                                (node->importing_count * sizeof(sds)));
+                            node->importing[node->importing_count - 2] =
+                                slot;
+                            node->importing[node->importing_count - 1] =
+                                src;
+                        }
+                    } else if ((dash = strchr(slotsdef, '-')) != NULL) {
+                        p = dash;
+                        int start, stop;
+                        *p = '\0';
+                        start = atoi(slotsdef);
+                        stop = atoi(p + 1);
+                        while (start <= stop) {
+                            int slot = start++;
+                            node->slots[node->slots_count++] = slot;
+                        }
+                    } else if (p > slotsdef) {
+                        int slot = atoi(slotsdef);
                         node->slots[node->slots_count++] = slot;
                     }
-                } else if (p > slotsdef) {
-                    int slot = atoi(slotsdef);
-                    node->slots[node->slots_count++] = slot;
                 }
             }
+            if (node->slots_count == 0) {
+                fprintf(stderr,
+                        "WARNING: Master node %s:%d has no slots, skipping...\n",
+                        node->ip, node->port);
+                continue;
+            }
+            if (!(config.cluster_nodes = addClusterNode(config.cluster_nodes, &config.cluster_node_count, node))) {
+                success = 0;
+                goto cleanup;
+            }
+            config.cluster_master_node_count++;
+            dictReplace(masters, node->name, node) ;
+        } else {
+            node->replicate = sdsnew(master_id);
+            
+            if (!(replica_nodes = addClusterNode(replica_nodes, &replica_count, node))) {
```

Comment:
The reason I added the replicas in a separate array and then added them at the end of the main cluster_nodes array is two fold. First, in the `CLUSTER NODES` response I believe a replica can come before a primary. So I wanted to ensure that the primary for a given replica had already been processed to determine what slots reside on the replica. Second and the main reason was I wanted to be able to support benchmarking both read and write commands with the `--readonly` flag without needing to restart redis-benchmark. In order to support this the easiest way was to put all the primary nodes at the front of the array. That way if the benchmarked command is a write command I just set the length of the cluster_nodes arrays as the number of primaries. That way the commands will only get sent to the primaries. If the command is a read command then set the length of the cluster_nodes array to include both primaries and replicas so that the commands get sent to both.

Reply:
I understand the motivation to place all masters at the beginning of the list. it is only that the implementation seems a bit complicated instead of a simple bubble sort loop on the nodes array after we add all cluster nodes...
IMO that way we can keep the code cleaner.

Reply:
The way I see there is a tradeoff between the current implementation and using a bubble sort to order the nodes at the end. With the current implementation some of the existing function prototypes need to be modified so that instead of modifying a global array you need to pass in the array as a parameter. Personally, I prefer this as it makes it clearer to me what the function is doing. On the other hand with a bubble sort the function prototypes can stay the same. 

However, since the order of `CLUSTER NODES` isn't guaranteed and replicas don't include the slots their primary owns the branch between replicas and primaries will still be required in the first for loop and the second for loop to update the slots for each replica will still be required. So I believe a bubble sort will result in more code. 

If you think a bubble sort is better I'll commit and make that change.

Reply:
@shawn-higgins1 - I do not mark that as a fundamental issue. I was just suggesting an alternative to avoid changing function prototypes. you can leave it as is if you feel that is clearer this way.